Validate incoming Record schemas

jeancochrane opened the pull request:
# Overview

In the course of reviewing #9, @ddohler noticed that we weren't actually validating that new Records matched their schema when creating them in the API; all of the validation logic was happening in the schema editor frontend.

This PR adds a validation step to Records to confirm that the new `data` field matches the associated RecordSchema.

## Notes

- Since the view tests and the Record validation code were both starting to get a bit bloated, I took this opportunity to do a small refactor: I split the Record validation tests out of `test_views.py` and into `test_record_validation.py`, and I reorganized `Record.clean()` to use descriptive methods (`clean_geom`, `clean_datetime`, `clean_data`). Apologies if this makes the diff harder to read!
- The commit history for this PR is not very clean, including one commit that introduces a failing test. As a result, I'll squash the PR before merging.

## Testing instructions

Travis will run the existing unit test suite for all supported versions of Python and Django. To run the tests locally:

```bash
# Build containers and run migrations.
./scripts/update

# Run tests.
./scripts/test app
```

Note that this will install virtualenvs for every version of Python and Django in the support matrix. If you'd like to remove virtualenvs after you're done, make sure to run `sudo rm -Rf .tox` in the project directory.

Closes https://github.com/azavea/grout-2018-fellowship/issues/44.

jeancochrane committed:
Move Record validation tests over into their own module

jeancochrane committed:
Add failing test for validating Records against their schema in the API

jeancochrane committed:
Validate the schema of incoming Records during the Record.clean() method

jeancochrane committed:
Update CHANGELOG and adjust schema validation test for python 2.7 error format

jeancochrane commented on tests/test_record_validators.py:
```diff
@@ -0,0 +1,139 @@
+        # The base endpoint for listing records.
+        cls.record_endpt = reverse('record-list')
+
+    def test_temporal_record_with_no_datetime_info(self):
```
This test was moved into this test module from `tests.test_views`.

jeancochrane commented on tests/test_record_validators.py:
```diff
@@ -0,0 +1,139 @@
+        self.assertEqual(min_msg, expected_msg, response.content)
+        self.assertEqual(max_msg, expected_msg, response.content)
+
+    def test_occurred_from_greater_than_occurred_to(self):
```
This test was also moved into this test module from `tests.test_views`.

ddohler commented on grout/models.py:
```diff
@@ -112,40 +113,52 @@ class Record(GroutModel):
-    def clean(self):
-        """
-        Provide custom field validation for this model.
+    def clean_geom(self):
```
Great refactor! 👍

ddohler reviewed (approved):
Looks good to me! My tests all passed locally but it might be a good idea to figure out why Travis is unhappy before merging.

jeancochrane committed:
Fix schema created in setup of JsonBFilterBackendTestCase

jeancochrane commented:
Looks like #10 introduced some test setup that didn't pass the new schema validation constraints. (Further verification that it works, at least!) I rebased off of develop and fixed the tests, so this is ready to roll.

Apologies for requesting your review before all tests passed, @ddohler! At the very least, this was a good lesson for me that I need to wait for both `pr` and `push` to pass before I can be confident all tests are passing.
